Angular feature toggles (#86)
Zuehlke/software-engineering-corner: drafts/angular-feature-toggles.md

publishAs: zemph
hideFromHashnodeCommunity: false
saveAsDraft: true
---

# Feature Flags in Angular

_Disclaimer: On a project I am currently working on, I introduced feature flags and created a setup in Angular that works great for me. 
With this post I would like to share the results with you._

## What are feature flags
In the most basic form, you can think of feature flags as a remote configuration consisting of features and their state (enabled / disabled). 
This configuration can be updated during the runtime of the application to  manage gradual rollouts, A/B tests, and quick rollbacks, improving flexibility and reducing risk.

<iframe src="https://stackblitz.com/edit/stackblitz-starters-6zh7dxms?embed=1&file=src%2Fmain.ts" width="100%" height="400"></iframe>

## Tooling

There are many tools available for adding feature flags to your project. 
The only requirement is a request, which returns a list of feature-flags with their current state. 
This could even be done by hosting a static JSON file somewhere that can easily be updated. 

## Goal

We are going to implement following three feature flags:
 - toggling the injection of an external analytics script into the DOM

>  - showing/hideing an iframe within our application

Reviewer: note:
in the example i tried to do it with an iframe. But it did not work. I changed the flag to a advertising banner. Idk maybe you want to change this here and in the example as well?

  }
};
```

### Fetching the feature flags
To toggle any kind of feature, we need to make the feature flag request as early as possible. 
Therefore we fetch the flags before bootstrapping the application. 
After receiving the remote feature flags, they can then pe provided within your Angular application using an injection token in your `ApplicationConfiguration`.

```js
// tokens.ts
// Create the injection token
export const FEATURE_FLAG_TOKEN = new InjectionToken<FeatureFlag>('FEATURE_FLAG_TOKEN');

// main.ts
// Fetch the feature flags and provide them within the ApplicationConfiguration

const fetchFlags = async (): Promise<FlagMap> => {
  try {
    // throw new Error('load default flags');
    const response = await fetch('/assets/feature-flags.json');
    return response.json();
  } catch {
    return DEFAULT_FLAGS;
  }
};

fetchFlags()
  .then((featureFlags) => bootstrapApplication(AppComponent, applicationConfig(featureFlags)));

// app.config.ts
export const applicationConfig = (featureFlags: FlagMap): ApplicationConfig => ({
  providers: [
    { provide: FEATURE_FLAG_TOKEN, useValue: featureFlags },
    // Rest of your providers
  ]
});
```

## Feature Flag Service

> In a service we can inject the remotely loaded feature flags and provide a function to check their status. 

Reviewer: suggestion:
i would add something before, to illustrate why we need a service.

This service will be used throughout our setup.

```js
@Injectable({ providedIn: 'root' })
export class FeatureFlagService {
  private readonly featureFlags = inject(FEATURE_FLAG_TOKEN);

  hasFeature(key: FlagKey): boolean {
    return this.featureFlags[key].enabled;
  }
}
```

### Basic use

With the basic setup established we are now able to use the feature flags service and check the flag values. 
We just need to inject it wherever we need it. 

```js

// Inject

> private readonly featureFlag = inject(FeatureFlagService);

Reviewer: suggestion:
i would call it featureFlagService :)

// Use
const featureValue = this.featureFlag.hasFeature('analytics')
```

More often than not, it’s not that simple to just see if a feature is enabled—we usually want to use feature flags in more specific contexts. 
Let me show you a few tools that can make your life easier when working with feature toggles.

## Using a feature flag before initializing Angular
To toggle any feature, it's essential to make the feature flag request as early as possible.
In Angular, there are different approaches to achieve this.
The earliest opportunity is to fetch the feature flags before bootstrapping the application.
For us this makes sense, since feature flags are fundamental to the app's core.

> Once resolved, the feature flags are provided through an InjectionToken.

Reviewer: question:
isnt this more a part of ## Fetching the feature flags?

Reviewer: imo the text does not fit the example

    if (featureFlag.hasFeature('analytics')) {
      addAnalyticsScriptToDom(document);
    }
  }),
},
```

## Conditionally showing UI using a directive
A very common use case for a feature flag is to show/hide content. 
This could be achieved with the code we already have but we can make our lives easier with creating a structural directive. 
It should take the flag key as an input and render/hide the component depending on the state of the flag 
Optionally it should also take a template reference to a fallback component, in case the feature is disabled.

```js
*appFeatureFlag="'iframe'; else fallback"
```

The directive has two inputs: one for the flag key and another for an optional fallback template. 
Based on the value of the flag, we either render the main template or the fallback template.

You might wonder why the input names seem weird.
This is necessary so the directive can be used as a structural directive with the `*` prefix.
This so called [structural directive shorthand](https://angular.dev/guide/directives/structural-directives#structural-directive-shorthand) requires us to use a certain naming convention for the inputs.

> Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that hosts the directive and surrounds the element and its descendants behind the scene. 

Reviewer: question:
why is this with the <ng-template> relevant again? i still dont understand why surrounding the element is important. Is it so we can add/remove it from the dom easily?

> Otherwise, we would have to do this ourselves anytime we use the directive. 

Reviewer: note:
i would move this part below the implementation.
You talk about input names, but they didnt see them yet

```js
@Directive({
  selector: '[appFeatureFlag]',
  standalone: true
})
export class FeatureFlagDirective {
  appFeatureFlag =  input<FlagKey | undefined>();
  appFeatureFlagElse = input<TemplateRef<unknown> | undefined>();

  private readonly featureFlag = inject(FeatureFlagService);
  private readonly templateRef = inject<TemplateRef<unknown>>(TemplateRef);
  private readonly viewContainer = inject(ViewContainerRef);

  private readonly isEnabled = computed(() => {
    const flagKey = this.appFeatureFlag();
    return !flagKey || this.featureFlag.hasFeature(flagKey);
  });

  constructor() {
    effect(() => {
      this.isEnabled() ? this.featureActive() : this.featureDisabled();
    });
  }

  private featureActive(): void {
    this.createView(this.templateRef);
  }

  private featureDisabled(): void {
    const elseTemplate = this.appFeatureFlagElse();
    elseTemplate ? this.createView(elseTemplate) : this.viewContainer.clear();
  }

  private createView(templateRef: TemplateRef<unknown>): void {
    this.viewContainer.clear();
    this.viewContainer.createEmbeddedView(templateRef);
  }
}
```

> And this is how you would use it within a template

Reviewer: ```suggestion
And this is how you would use it within a template.
```

```html
<iframe *appFeatureFlag="'iframe'; else fallback"
        width="560"
        height="315"
        src="https://www.youtube.com/embed/dQw4w9WgXcQ?si=PqWzCGoitRdn2RUt"
        title="YouTube video player" referrerpolicy="strict-origin-when-cross-origin"></iframe>
<ng-template #fallback>
    Content unavailable :/
</ng-template>
```

## Guarding a route with a feature flag

Protecting a route with a feature flag guard can be quite useful for restricting access to certain parts of your application. 
If the feature is active, the guard returns true and allows the navigation. 
Otherwise it redirects to the fallback url.

```js
export function featureFlagGuard(key: FlagKey, redirectUrl: string = '/'): CanActivateFn {
  return () => {
    const featureFlag = inject(FeatureFlagService);
    const router = inject(Router);

    if (featureFlag.hasFeature(key)) {
      return true;
    }
    if (redirectUrl) {
      return router.parseUrl(redirectUrl);
    }

>     return router.parseUrl(HOME_PAGE_URL);

Reviewer: note:
this is not needed anymore, because redirectUrl is always available

  };
}
```

This `canActivate` guard can be used with any route.

```js
{
  path: 'new-feature',
  component: NewFeatureComponent,
  pathMatch: 'full',
  canActivate: [featureFlagGuard('route', '/coming-soon')],
}
```

## Conclusion
While feature flags can add flexibility and support gradual rollouts or A/B testing, they also bring extra complexity—and, if not handled carefully, can cause issues. 

> But when used with clear guidelines, they can help teams experiment and adapt more easily without constant redeployments. 

Reviewer: suggestion:
maybe add a: let me know about your experiences with feature flags, comment ....

> ---

Author: Not all of them are

title: Feature Toggles in Angular
subtitle: "Crafting a Custom Feature Toggle Setup in Angular"
domain: software-engineering-corner.zuehlke.com
tags: angular, web-development, javascript, frontend-development, guide
cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1732888744339/n2T03yjCA.jpg?auto=format
publishAs: zemph
hideFromHashnodeCommunity: false
saveAsDraft: true
---

_Disclaimer: This post doesn’t aim to promote feature toggles, but rather to demonstrate what a custom setup in Angular could look like. 
When used appropriately, feature toggles can be a powerful tool. 
However, if overused or poorly managed, they can introduce unnecessary complexity to your project._

> ## Tooling

Author: I removed the sentence about flagsmith to make it more general, what the tooling requirements for angular are.

    },
    {
        "feature": {
            "id": 1,
            "name": "iframe"
        },
        "enabled": true
    },
    {
        "feature": {
            "id": 2,
            "name": "route"
        },
        "enabled": true
    }
]
```

## Basic Setup

### Type Safety

For some type safety within the feature toggle setup, I came up with the following model.

```js
// feature-toggle.model.ts
import { IFlagsmith } from 'flagsmith/types';

export type FeatureToggle = IFlagsmith<FlagKey, TraitKey>;

export type FlagKey = 'analytics' | 'iframe' | 'route';

export interface Flag {
  readonly key: FlagKey;
  readonly enabled: boolean;
}

export type FlagMap = {
  readonly [key in FlagKey]: Flag;
};
```

Using the `FlagMap` type, we can create a constant to manage and access all our flags and their keys. 
We can also use it to set our default flags, which we needed when initiating the feature toggle instance.

```js
// feature-toggle.constants.ts
export const featureFlags: FlagMap = {
  analytics: {
    key: 'analytics',
    enabled: true
  },
  iframe: {
    key: 'iframe',
    enabled: false
  },
  route: {
    key: 'route',
    enabled: true
  }
};
```

> ### Setting up the feature toggle instance

Reviewer: ```suggestion
### Fetching the feature flags
```
See how I did it in the Stackblitz. in the end you fetch the feature flags in a similar way, the only difference is you dont pass the flagsmith instance, but the flags itself.
Then you use a injection token to make them available. Then there is a feature flag service, which consums these flags. Super easy and the focus is on your angular magic

> To toggle any kind of feature, we need to make the feature flag request as early as possible. 
> Therefore we create our feature toggle instance and make the initial call, which fetches the flags. 
> We do this before bootstrapping the application. 
> The feature toggle instance should provide some method to check if a feature is active at any given time, without having to re-fetch the flags. 
> With Flagsmith its the `hasFeature()` function. 
> After setting up the feature toggle instance, it can then pe provided within your Angular application using an injection token in your `ApplicationConfiguration`.

Reviewer: ```suggestion
To toggle any feature, it's essential to make the feature flag request as early as possible.
In Angular, there are different approaches to achieve this.
The earliest opportunity is to fetch the feature flags before bootstrapping the application.
For us this makes sense, since feature flags are fundamental to the app's core.
Once resolved, the feature flags are provided through an InjectionToken.
```

```js
// tokens.ts
// Create the injection token
export const FEATURE_TOGGLE_TOKEN = new InjectionToken<FeatureToggle>('FEATURE_TOGGLE_TOKEN');

// main.ts
// Create feature toggle instance and provide within the ApplicationConfiguration
const flagsmith = createFlagsmithInstance();
flagsmith
  .init({
    environmentID: ENVIRONMENT_ID,
    defaultFlags: DEFAULT_FLAGS
  })
  .then(() => bootstrapApplication(AppComponent, applicationConfig(flagsmith)));

// app.config.ts
export const applicationConfig = (flagsmith: FeatureToggle): ApplicationConfig => ({
  providers: [
    { provide: FEATURE_TOGGLE_TOKEN, useValue: flagsmith },
    // Rest of your providers
  ]
});
```

## Basic use

With the basic setup established we are now able to use the feature toggle instance and check the flag values. 
We just need to inject it wherever we need it. 

```js

// Inject
private readonly featureToggle = inject(FEATURE_TOGGLE_TOKEN);

// Use
const featureValue = this.featureToggle.hasFeature('analytics')
```

This flag’s value can now be used to change the behavior of your application. 
More often than not, it’s not that simple—we usually want to use feature flags in more specific contexts. 

> Thats why I want to talk about a few tools that can make your life easier when working with feature toggles.

Reviewer: suggestion:
you might want to describe the feature flag service here

> This flag’s value can now be used to change the behavior of your application. 
> More often than not, it’s not that simple—we usually want to use feature flags in more specific contexts. 
> Thats why I want to talk about a few tools that can make your life easier when working with feature toggles.

Reviewer: ```suggestion
More often than not, it’s not that simple to just see if a feature is enabled—we usually want to use feature flags in more specific contexts. 
Let me show you a few tools that can make your life easier when working with feature toggles.
```

## Using a feature flag before initializing Angular

> The first situation I want to talk about is using a feature flag, before the application is fully initialized. 

Reviewer: note:
I think this sentence is not necessary, since it just repeats the title.

In this example we want to conditionally inject an analytics script into the DOM to track the user interaction within our application. 
To achieve this, we can use the `provideAppInitializer` function within our `ApplicationConfiguration`. 
Using our feature toggle instance, we can conditionally append the script to the DOM depending on the flag.

```js
// app.config.ts
{
    provideAppInitializer(() => {
    const featureToggle = inject(FEATURE_TOGGLE_TOKEN);

    if (featureToggle.hasFeature('analytics')) {
      addAnalyticsScriptToDom(document);
    }
  }),
},
```

## Conditionally showing UI using a directive
A very common use case for a feature flag is to show/hide content. 

> This could be achieved with the setup we already have but we can make our lives easier when creating a directive we can use within the template. 

Reviewer: ```suggestion
This could already be achieved with the code we already have but we can make our lives easier with creating a structural directive. 
```

> It should take the flag key as an input and render/hide the component depending on the flag value. 

Reviewer: ```suggestion
It should take the flag key as an input and render/hide the component depending on the state of the flag 
```

> Optionally it should also take a template reference as an input for the fallback, if the feature is disabled.

Reviewer: ```suggestion
Optionally it should also take a template reference to a fallback component, in case the feature is disabled.
```

> 

Reviewer: suggestion:
add here just a quick preview of the directive and how it should work:
`*featureToggle="'iframe'; else fallback"`

> The setup of the feature toggle directive is quite simple. 
> We inject the feature toggle instance, the `TemplateRef`, and the `ViewContainerRef`. 
> We define two inputs: one for the flag key and another for an optional fallback template. 
> A `computed` signal checks if the feature is enabled. 
> Inside the constructor, we use an `effect` to reactively respond to changes in the signal. 
> Based on the value, we either render the main template using `featureActive()` or the fallback template using `featureDisabled()`.

Reviewer: ```suggestion
The directive has two inputs: one for the flag key and another for an optional fallback template. 
Based on the value of the flag, we either render the main template or the fallback template.
```

Reviewer: imo the code is quite simple, so idk if it makes sense the explain everything

If the feature is active, we use the `ViewContainerRef` to create an embedded view from the `TemplateRef` that the directive is attached to. 
If the feature is not active, we clear the view container and, if a fallback template is provided via the `appFeatureToggleElse` input, render that instead.

> 

Reviewer: suggestion:
either leave this out or explain better what ViewContainerRef and TemplateRef do

> You might wonder why the input names seem weird. 
> They are set like this on purpose to, so we can use this directive as a structural directive with the `*` prefix. 
> This is beneficial because Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that hosts the directive and surrounds the element and its descendants behind the scene. 
> Otherwise, we would have to do this ourselves anytime we use the directive. 
> This so called [structural directive shorthand](https://angular.dev/guide/directives/structural-directives#structural-directive-shorthand) requires us to use a certain naming convention for the inputs. 
> They need to have the directive selector as a prefix. 

Reviewer: ```suggestion
You might wonder why the input names seem weird.
This is necessary so the directive can be used as a structural directive with the `*` prefix.
This so called [structural directive shorthand](https://angular.dev/guide/directives/structural-directives#structural-directive-shorthand) requires us to use a certain naming convention for the inputs.
This is beneficial because Angular transforms the asterisk in front of a structural directive into an `<ng-template>` that hosts the directive and surrounds the element and its descendants behind the scene. 
Otherwise, we would have to do this ourselves anytime we use the directive. 
```

question:
why is it benefitial that there is a ng-template wrapping happening?

```js
@Directive({
  selector: '[appFeatureToggle]',
  standalone: true
})
export class FeatureToggleDirective {
  appFeatureToggle =  input<FlagKey | undefined>();
  appFeatureToggleElse = input<TemplateRef<unknown> | undefined>();

  private readonly featureToggle = inject(FEATURE_TOGGLE_TOKEN);
  private readonly templateRef = inject<TemplateRef<unknown>>(TemplateRef);
  private readonly viewContainer = inject(ViewContainerRef);

  private readonly isEnabled = computed(() => {
    const flagKey = this.appFeatureToggle();
    return !flagKey || this.featureToggle.hasFeature(flagKey);
  });

  constructor() {
    effect(() => {
      this.isEnabled() ? this.featureActive() : this.featureDisabled();
    });
  }

  private featureActive(): void {
    this.createView(this.templateRef);
  }

  private featureDisabled(): void {
    const elseTemplate = this.appFeatureToggleElse();
    elseTemplate ? this.createView(elseTemplate) : this.viewContainer.clear();
  }

  private createView(templateRef: TemplateRef<unknown>): void {
    this.viewContainer.clear();
    this.viewContainer.createEmbeddedView(templateRef);
  }
}
```

And this is how you would use it within a template

```html

> <iframe *featureToggle="'iframe'; else fallback"

Reviewer: ```suggestion
<iframe *appFeatureToggle="'iframe'; else fallback"
```

        width="560"
        height="315"
        src="https://www.youtube.com/embed/dQw4w9WgXcQ?si=PqWzCGoitRdn2RUt"
        title="YouTube video player" referrerpolicy="strict-origin-when-cross-origin"></iframe>
<ng-template #fallback>
    Content unavailable :/
</ng-template>
```

## Guarding a route with a feature flag

> In this last example I want to demonstrate how you can use a guard to toggle a route in your Angular project. 

Reviewer: remove this line, title says it all

> It’s easy to implement and can be quite useful for restricting access to certain parts of your application. 

Reviewer: ```suggestion
Protecting a route with a feature flag guard can be quite useful for restricting access to certain parts of your application. 
```

If the feature is active, the guard returns true and allows the navigation. 

> Otherwise it redirects to the fallback url if provided. 
> In my case the fallback is optional and the guard navigates to the home page by default.

Reviewer: ```suggestion
Otherwise it redirects to the fallback url.
```

```js

> export function featureToggleGuard(key: FlagKey, redirectUrl?: string): CanActivateFn {

Reviewer: ```suggestion
export function featureToggleGuard(key: FlagKey, redirectUrl: string = '/'): CanActivateFn {
```

  return () => {
    const featureToggle = inject(FEATURE_TOGGLE_TOKEN);
    const router = inject(Router);

    if (featureToggle.hasFeature(key)) {
      return true;
    }
    if (redirectUrl) {
      return router.parseUrl(redirectUrl);
    }
    return router.parseUrl(HOME_PAGE_URL);
  };
}
```

> This guard can be used in any route within the `canActivate` Array.

Reviewer: ```suggestion
This `canActivate` guard can be used with any route.
```

```js
{
  path: 'new-feature',
  component: NewFeatureComponent,
  pathMatch: 'full',
  canActivate: [featureToggleGuard('route', '/coming-soon')],
}
```

## Conclusion

> Feature flags aren’t the right choice for every project. 

Reviewer: question:
why? elaborate

While they can add flexibility and support gradual rollouts or A/B testing, they also bring extra complexity—and, if not handled carefully, can cause issues. But when used with clear guidelines, they can help teams experiment and adapt more easily.

> In this post, I shared a custom setup for feature toggles in Angular using Flagsmith, including examples for runtime configuration, conditionally showing UI with directives, and guarding routes. 

Reviewer: qustion:
is this necessary?

Author: Updated it
